pygen/pygen_src/riscv_asm_program_gen.py

```diff
@@ -437,15 +437,35 @@ def init_floating_point_gpr_with_dpf(self, int_floating_gpr):
         self.instr_stream.extend((li_instr0, slli_instr, li_instr1, or_instr, fmv_instr))
 
     # Get a random single precision floating value
+    @staticmethod
+    def get_randselect0(addr_range1, addr_range2):
+        value = vsc.rand_bit_t(32)
+        with vsc.randomize_with(value): value in vsc.rangelist(addr_range1, addr_range2)
+
     def get_rand_spf_value(self):
-        # TODO randcase
-        value = random.randrange(0, 2**32 - 1)
+        vsc.randselect([
+            (1, lambda: get_randselect0(hex(0x7f80_0000), hex(0xff80_0000))),
+            (1, lambda: get_randselect0(hex(0x7f7f_ffff), hex(0xff7f_ffff))),
+            (1, lambda: get_randselect0(hex(0x0000_0000), hex(0x8000_0000))),
+            (1, lambda: get_randselect0(hex(0x7f80_0001), hex(0x7fc0_0000))),
+            (1, lambda: vsc.rng(30, pkg_ins.SINGLE_PRECISION_FRACTION_BITS) > 0)
+            (1, lambda: vsc.rng(30, pkg_ins.SINGLE_PRECISION_FRACTION_BITS) == 0)])
         return value
 
     # Get a random double precision floating value
-    def get_rand_dpf_value(self):
+    @staticmethod
+    def get_randselect1(addr_range1, addr_range2):
         value = vsc.bit_t(64)
-        # TODO randcase
+        with vsc.randomize_with(value): value in vsc.rangelist(addr_range1, addr_range2)
+
+    def get_rand_dpf_value(self):
+        vsc.randselect([
+            (1, lambda: get_randselect1(hex(0x7ff0_0000_0000_0000), hex(0xfff0_0000_0000_0000))),
```

Comment:
same here (re: adding comments about the address ranges), linking to the appropriate area of the spec would work as well IMO